### PR TITLE
Prototype/starport

### PR DESCRIPTION
**Pull request by syan095**
First prototype for the Starport Module
* Contains Extrinsics call for Locking of assets and Invoking notices from Compound Gateway.
* Added an new place holder module for Compound Cash
* Integrated into Mandala Runtime.

### CODE REVIEW

File: modules/starport/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+[features]
+default = ["std"]
+std = [
+	"serde",
+	"codec/std",
+	"frame-support/std",
+	"frame-system/std",
+	"sp-runtime/std",
+	"sp-core/std",
+    "sp-io/std",
```
Comment (xlc) [inline modules/starport/Cargo.toml:32]: use tab

File: modules/starport/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+			// After its invocation, store the hash.
+			InvokedNoticeHashes::<T>::insert(&hash, ());
+
+			Ok(().into())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	#[transactional]
```
Comment (xlc) [inline modules/starport/src/lib.rs:295]: no need to put `#[transactional]` in private inner help method. could use `#[require_transactional]`

File: modules/starport/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+		from: T::AccountId,
+		to: T::AccountId,
+		currency_id: CurrencyId,
+		locked_amount: Balance,
+	) -> DispatchResultWithPostInfo {
+		// Ensure the user has sufficient balance
+		ensure!(
+			T::Currency::ensure_can_withdraw(currency_id.clone(), &from, locked_amount.clone()).is_ok(),
+			Error::<T>::InsufficientBalance
+		);
```
Comment (xlc) [inline modules/starport/src/lib.rs:306]:
```suggestion
		T::Currency::ensure_can_withdraw(currency_id.clone(), &from, locked_amount.clone())?;
```

**Comment by wangjj9219** [comment]
maybe better to put it in the `ecosystem-modules` folder

File: ecosystem-modules/starport/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Multi-currency support for asset management
+		type Currency: MultiCurrency<Self::AccountId, CurrencyId = CurrencyId, Balance = Balance>;
+
+		/// The Admin account that executes Notices from Compound chain
```
Comment (zjb0807) [inline ecosystem-modules/starport/src/lib.rs:68]: add `#[pallet::constant]` and check others.

File: ecosystem-modules/starport/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+		/// Update the current set of authorities who sign Notices.
+		ChangeAuthorities(Vec<CompoundAuthoritySignature>),
+
+		/// Unlock or download assets from Compound chain back into Acala chain.
+		Unlock(CurrencyId, Balance, AccountId),
+
+		/// Set the future yield for the Cash asset.
+		/// Parameters: uint128 nextCashYield, uint128 nextCashYieldIndex, uint nextCashYieldStart
+		SetFutureYield(Balance, CashYieldIndex, Moment),
```
Comment (xlc) [inline ecosystem-modules/starport/src/lib.rs:161]:
```suggestion
		SetFutureYield { next_cash_yield: Balance, next_cash_yield_index: CashYieldIndex, next_cash_yield_start: Moment },
```

Prefer to use named arguments if there are multiple arguments

File: ecosystem-modules/starport/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+	/// Stores the amount of supplies that are still available to be uploaded for each asset type.
+	#[pallet::storage]
+	#[pallet::getter(fn supply_caps)]
+	pub type SupplyCaps<T: Config> = StorageMap<_, Twox64Concat, CurrencyId, Balance, ValueQuery>;
+
+	/// Stores the Hash of Notices that have already been invoked. PrevenTwox64Concatts
+	/// double-invocation.
+	#[pallet::storage]
+	#[pallet::getter(fn invoked_notice_hashes)]
+	pub type InvokedNoticeHashes<T: Config> = StorageMap<_, Identity, H256, (), ValueQuery>;
```
Comment (xlc) [inline ecosystem-modules/starport/src/lib.rs:173]:
```suggestion
	pub type InvokedNoticeHashes<T: Config> = StorageMap<_, Identity, H256, (), OptionQuery>;
```

File: ecosystem-modules/compound-cash/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	pub type CurrentYield<T: Config> = StorageValue<_, (CashYieldIndex, Balance, Moment), ValueQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			// check if the Yield rate needs to be updated
+			0
```
Comment (shaunxw) [inline ecosystem-modules/compound-cash/src/lib.rs:84]: Is it to be done in this PR?

File: ecosystem-modules/compound-cash/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {}
```
Comment (shaunxw) [inline ecosystem-modules/compound-cash/src/lib.rs:87]: No need to include `on_finalize` if no custom impl.

File: ecosystem-modules/compound-cash/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+
+use frame_support::{pallet_prelude::*, traits::UnixTime};
+use module_support::CompoundCashTrait;
+use primitives::{Balance, CashYieldIndex, CurrencyId, Moment, TokenSymbol};
+
+// mod mock;
+// mod tests;
```
Comment (shaunxw) [inline ecosystem-modules/compound-cash/src/lib.rs:32]:
```suggestion
mod mock;
mod tests;
```

File: ecosystem-modules/compound-cash/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn set_future_yield(
+		next_cash_yield: Balance,
+		yield_index: CashYieldIndex,
+		timestamp_effective: Moment,
+	) -> DispatchResultWithPostInfo {
```
Comment (shaunxw) [inline ecosystem-modules/compound-cash/src/lib.rs:99]: Use `DispatchResult` if not returning post info. Same to all other Pallet fn and dispatchable calls.

File: ecosystem-modules/starport/src/lib.rs (hunk 6 above)
Comment (xlc) [reply ecosystem-modules/starport/src/lib.rs:173]: also check if OptionQuery should be used for other storages. we should default o use OptionQuery except for the case missing value can be same as default value

File: ecosystem-modules/compound-cash/src/tests.rs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+//! Unit tests for example module.
+
+#![cfg(test)]
+
+use crate::mock::*;
+use frame_support::assert_ok;
+
+#[test]
+fn set_dummy_work() {
```
Comment (shaunxw) [inline ecosystem-modules/compound-cash/src/tests.rs:27]: We need some real unit tests.

File: ecosystem-modules/starport/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+		ensure!(
+			current_supply_cap >= locked_amount,
+			Error::<T>::InsufficientAssetSupplyCap
+		);
+
+		// If the currency is CASH, it is burned
+		// All other tokens are transferred to the admin's account.
+		match currency_id {
+			CurrencyId::Token(TokenSymbol::CASH) => T::Currency::withdraw(currency_id, &from, locked_amount),
+			_ => T::Currency::transfer(currency_id, &from, &T::AdminAccount::get(), locked_amount),
```
Comment (xlc) [inline ecosystem-modules/starport/src/lib.rs:385]: should just replace admin account to `PalletId::get().into_account(),`

File: ecosystem-modules/compound-cash/src/lib.rs (hunk 7 above)
Comment (syan095) [reply ecosystem-modules/compound-cash/src/lib.rs]: It's just a empty place holder for now, will be completed once the spec is confirmed

File: ecosystem-modules/compound-cash/src/tests.rs (hunk 11 above)
Comment (syan095) [reply ecosystem-modules/compound-cash/src/tests.rs]: Removed the dummy tests. Actual tests will be added once the module is spec is locked and implemented

**Review by xlc — APPROVED** [review]

**Pull request merged**